token-lending/program/tests/helpers/flash_loan_receiver.rs

```diff
@@ -0,0 +1,147 @@
+    ///
+    /// Accounts expected:
+    ///
+    /// 0. `[writable]` The source liquidity token account.
+    /// 1. `[writable]` The destination liquidity token account (i.e. reserve liquidity token account).
+    /// 2. `[]` The token program
+    /// 3. `[]` program derived account.
+    ReceiveFlashLoan {
+        /// The amount that is loaned
+        amount: u64,
+    },
+}
+
+entrypoint!(process_instruction);
+pub fn process_instruction(
+    program_id: &Pubkey,
+    accounts: &[AccountInfo],
+    instruction_data: &[u8],
+) -> ProgramResult {
+    Processor::process(program_id, accounts, instruction_data)
+}
+
+pub struct Processor;
+impl Processor {
+    pub fn process(
+        program_id: &Pubkey,
+        accounts: &[AccountInfo],
+        instruction_data: &[u8],
+    ) -> ProgramResult {
+        let instruction = FlashLoanReceiverInstruction::unpack(instruction_data)?;
+
+        match instruction {
+            FlashLoanReceiverInstruction::ReceiveFlashLoan { amount } => {
+                msg!("Instruction: Receive Flash Loan");
+                Self::process_receive_flash_loan(accounts, amount, program_id)
+            }
+        }
+    }
+
+    fn process_receive_flash_loan(
+        accounts: &[AccountInfo],
+        amount: u64,
+        program_id: &Pubkey,
+    ) -> ProgramResult {
+        let account_info_iter = &mut accounts.iter();
+        let source_liquidity_token_account_info = next_account_info(account_info_iter)?;
+        let destination_liquidity_token_account_info = next_account_info(account_info_iter)?;
+        let token_program_id = next_account_info(account_info_iter)?;
+        let program_derived_account_info = next_account_info(account_info_iter)?;
+
+        let destination_liquidity_token_account = Account::unpack_from_slice(
+            &source_liquidity_token_account_info.try_borrow_mut_data()?,
+        )?;
+        let (expected_program_derived_account_pubkey, bump_seed) =
+            Pubkey::find_program_address(&[b"flashloan"], program_id);
+
+        if &expected_program_derived_account_pubkey != program_derived_account_info.key {
+            msg!("Supplied program derived account doesn't match with expectation.")
+        }
+
+        if destination_liquidity_token_account.owner != expected_program_derived_account_pubkey {
+            msg!("Destination liquidity token account is not owned by the program");
+            return Err(ProgramError::IncorrectProgramId);
+        }
+
+        let balance_in_token_account =
+            Account::unpack_from_slice(&source_liquidity_token_account_info.try_borrow_data()?)?
+                .amount;
+        let transfer_ix = spl_token::instruction::transfer(
+            token_program_id.key,
+            source_liquidity_token_account_info.key,
+            destination_liquidity_token_account_info.key,
+            &expected_program_derived_account_pubkey,
+            &[],
+            min(balance_in_token_account, amount),
+        )?;
```

Comment:
Passing something like `flash_loan_amount` and `flash_loan_fee`, or `flash_loan_amount` and `repay_amount`, will make this nicer to work with for end users instead of having to intuit how much to use.

Reply:
Just to clarify, are you suggesting that we change the signature of `ReceiveFlashLoan` from accepting one argument into accepting two, i.e. go from `ReceiveFlashLoan {amount: u64}` -> `ReceiveFlashLoan {flash_loan_amount: u64, flash_loan_fee:u64}`?

Reply:
The logic here is mainly to accommodate the `test_failure`. Normally, when user have enough fund the user will only need to transfer the amount that passed into the `flash_loan_receiver_program` as in the previous commit. However, in our case we deliberately reduce the token amount by 1 to failed the test, so the `flash_loan_receiver_program` needs to check how much it actually have and then only return that amount.

Reply:
Yeah, this note is a duplicate of https://github.com/solana-labs/solana-program-library/pull/1444#discussion_r625413753 with more context there.

Reply:
I see! Thanks for clarifying!